File: packages/go/internal/seguranca/ataques/ataque_ida_e_volta_binario.go
package ataques

import (
	"fmt"
	"strings"

	"criptografia/internal/seguranca"
)

// AtaqueIdaEVoltaBinario é a versão binária do AtaqueIdaEVolta: usa todos os
// 256 valores de byte, NUL e sequências aleatórias de vários tamanhos. Erros de
// manipulação de string (trim, encoding, truncamento em NUL) só aparecem com
// bytes arbitrários.
type AtaqueIdaEVoltaBinario struct {
	tamanhoMaximo int
}

// NovoAtaqueIdaEVoltaBinario cria o ataque. Sem argumento, usa
// tamanhoMaximo = 256.
func NovoAtaqueIdaEVoltaBinario(tamanhoMaximo ...int) *AtaqueIdaEVoltaBinario {
	t := 256
	if len(tamanhoMaximo) > 0 {
		t = tamanhoMaximo[0]
	}
	return &AtaqueIdaEVoltaBinario{tamanhoMaximo: t}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueIdaEVoltaBinario) Nome() string {
	return "Ida-e-volta com dados binários (inclui NUL)"
}

// Executar roda o ataque.
func (a *AtaqueIdaEVoltaBinario) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	falhas := []string{}

	for length := 0; length <= a.tamanhoMaximo; length++ {
		texto := []byte{}
		if length > 0 {
			texto = seguranca.RandomBytes(length)
		}
		token, err := alvo.Encrypt(texto)
		if err != nil {
			falhas = append(falhas, fmt.Sprintf("%d (erro: %s)", length, err.Error()))
			continue
		}
		decifrado, err := alvo.Decrypt(token)
		if err != nil {
			falhas = append(falhas, fmt.Sprintf("%d (erro: %s)", length, err.Error()))
			continue
		}
		if decifrado != string(texto) {
			falhas = append(falhas, fmt.Sprintf("%d", length))
		}
	}

	todos := make([]byte, 256)
	for i := 0; i < 256; i++ {
		todos[i] = byte(i)
	}
	tokenTodos, err := alvo.Encrypt(todos)
	if err != nil {
		falhas = append(falhas, "todos os 256 valores de byte")
	} else if decifrado, err := alvo.Decrypt(tokenTodos); err != nil || decifrado != string(todos) {
		falhas = append(falhas, "todos os 256 valores de byte")
	}

	if len(falhas) > 0 {
		limite := min(10, len(falhas))
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevCritica,
			Detalhes:   "Falhou em " + fmt.Sprint(len(falhas)) + " caso(s): " + strings.Join(falhas[:limite], ", "),
			Dados:      map[string]any{"falhas": falhas},
		}, nil
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: false,
		Severidade: seguranca.SevInfo,
		Detalhes:   fmt.Sprintf("Todos os tamanhos 0..%d e os 256 valores de byte preservados", a.tamanhoMaximo),
	}, nil
}

File: packages/python/src/Seguranca/Ataques/AtaqueIdaEVoltaBinario.py
"""O AtaqueIdaEVolta usa só o caractere 'Q'. Esse usa dados binários de
verdade: todos os 256 valores de byte, NUL, sequências aleatórias de
vários tamanhos. Erros de manipulação de string (trim, encoding, NUL
truncation) só aparecem com bytes arbitrários.

Adaptação Python: a API do alvo recebe/devolve `str` (codificado em UTF-8
internamente), então mapeamos cada byte para um caractere via latin-1 -
uma bijeção byte<->code point, que preserva NUL e todos os 256 valores -
e reencodamos o resultado para comparar os BYTES originais.
"""

from ..ResultadoAtaque import ResultadoAtaque
from ..Util import random_bytes


class AtaqueIdaEVoltaBinario:
    def __init__(self, tamanho_maximo: int = 256):
        self.tamanho_maximo = tamanho_maximo

    def nome(self) -> str:
        return "Ida-e-volta com dados binários (inclui NUL)"

    def executar(self, alvo) -> ResultadoAtaque:
        falhas = []

        for length in range(self.tamanho_maximo + 1):
            texto_bytes = b"" if length == 0 else random_bytes(length)
            texto = texto_bytes.decode("latin-1")
            try:
                decifrado = alvo.decrypt(alvo.encrypt(texto))
                if decifrado.encode("latin-1") != texto_bytes:
                    falhas.append(length)
            except Exception as e:
                falhas.append(f"{length} (erro: {e})")

        todos_bytes = bytes(range(256))
        todos = todos_bytes.decode("latin-1")
        if alvo.decrypt(alvo.encrypt(todos)).encode("latin-1") != todos_bytes:
            falhas.append("todos os 256 valores de byte")

        if falhas:
            return ResultadoAtaque(
                self.nome(),
                True,
                "critica",
                "Falhou em "
                + str(len(falhas))
                + " caso(s): "
                + ", ".join(str(f) for f in falhas[:10]),
                {"falhas": falhas},
            )

        return ResultadoAtaque(
            self.nome(),
            False,
            "info",
            f"Todos os tamanhos 0..{self.tamanho_maximo} e os 256 valores "
            "de byte preservados",
        )
